Launcher/src/windows/update.c

```diff
@@ -0,0 +1,116 @@
+#include <stdio.h>
+#include <sys/stat.h>
+#include <windows.h>
+
+#define BUFFER_SIZE (1024 * 1024 * 16)
+#define DELIM "\\"
+
+int copyData(struct archive *in, struct archive *out) {
+	const void *buff;
+	size_t size;
+	la_int64_t offset;
+	while (1) {
+		int result = archive_read_data_block(in, &buff, &size, &offset);
+		if (result == ARCHIVE_EOF)
+			return ARCHIVE_OK;
+		else if (result < ARCHIVE_OK)
+			return result;
+		result = archive_write_data_block(out, buff, size, offset);
+		if (result < ARCHIVE_OK) {
+			return result;
+		}
+	}
+	return ARCHIVE_OK;
+}
+
+void launcher_update(char* baseDir) {
+	assert(baseDir);
+	char *updateDir = malloc(strlen(baseDir) + 3 + 9);
+	assert(updateDir);
+	sprintf(updateDir, "%s%s%s", baseDir, DELIM, "updates");
+	char *updateFile = malloc(strlen(updateDir) + 3 + 10);
+	assert(updateFile);
+	sprintf(updateFile, "%s%s%s", updateDir, DELIM, "data.zip");
+
+	struct stat s;
+	if (stat(updateFile, &s) != 0) {
+		printf("No update-file found in %s\n", updateDir);
+		free(baseDir);
+		free(updateDir);
+		free(updateFile);
+		return;
+	}
+	printf("Update found\n");
+
+	int flags = ARCHIVE_EXTRACT_TIME;
+	flags |= ARCHIVE_EXTRACT_PERM;
+	flags |= ARCHIVE_EXTRACT_ACL;
+	flags |= ARCHIVE_EXTRACT_FFLAGS;
+
+	struct archive *a = archive_read_new();
+	struct archive *out = archive_write_disk_new();
+	archive_write_disk_set_options(out, flags);
+	archive_write_disk_set_standard_lookup(out);
+	struct archive_entry *entry;
+	int r;
+
+	archive_read_support_compression_all(a);
+	archive_read_support_format_zip(a);
+	if (r = archive_read_open_filename(a, updateFile, BUFFER_SIZE)) {
+		fprintf(stderr, "Couldn't open %s: %s\n", updateFile,
+				archive_error_string(a));
+		exit(-1);
+	}
+	printf("Extracting archive...\n");
+	for (;;) {
+		r = archive_read_next_header(a, &entry);
+		if (r == ARCHIVE_EOF) {
+			break;
+		}
+		char *cc = calloc(
+			strlen(baseDir) + strlen(archive_entry_pathname(entry)) + 1, 1);
+		assert(cc);
+		sprintf(cc, "%s%s%s", baseDir, DELIM, archive_entry_pathname(entry));
+		archive_entry_set_pathname(entry, cc);
+		r = archive_write_header(out, entry);
+		if (r != ARCHIVE_OK) {
+			fprintf(stderr, "Error writing header %s: %d %s",
+					archive_entry_pathname(entry), r, archive_error_string(a));
+			exit(-1);
+		}
+		if (archive_entry_size(entry) > 0) {
+			r = copyData(a, out);
+			if (r != ARCHIVE_OK) {
+				fprintf(stderr, "Error while copying data: %d\n", r);
+				exit(-1);
+			}
+		}
+		r = archive_write_finish_entry(out);
+		if (r != ARCHIVE_OK) {
+			fprintf(stderr, "Error finishing entry: %d %s\n", r,
+					archive_error_string(a));
+			exit(-1);
+		}
+		free(cc);
+	}
+	archive_read_close(a);
+	archive_read_free(a);
+	
+	remove(updateFile);
+	
+	char* installer = malloc(strlen(updateDir) + 3 + 14);
+	sprintf(installer, "%s%s%s", updateDir, DELIM, "Installer.exe");
+	
+	printf("Starting Installer: %d\n", ShellExecute(0, TEXT("runas"), TEXT(installer), 0, NULL, SW_HIDE));
```

Comment:
I frankly don't like this line. I was thinking of two ways of updating:

1. Download the datazip. Just extract it over %PROGRAMFILES%/Conquer. Pro: It is simple; Contra: A lot of bloat may be created over the times.
2. Download the datazip and make a standardised package format (E.g. an installation "script" in META-INF/someFile.json with the files to replace, to add, to delete and to rename). Pro: No bloat and more flexible. Contra: Complicated

Another issue with this line is, that you can't delete/change open .exe files in windows. If you start the installer, the launcher is still running => The installer will just fail silently.

Reply:
I'm going to give you free hand, but I'm advocating for (2), as it is more flexible in a long way.

The issue of the open files could be avoided, if the program flow is like this:
1.data.zip is downloaded. It contains updated dlls, updated launcher, updated jars, ...
2.You extract it on Java side to e.g. `%APPDATA%\.conquer\tmp`.
3.You "compile" the setup script to a batch file with computer-specific paths. It will look like this:
```
@echo off
:REM: Sleep here for some time
move %APPDATA%\.conquer\tmp\foo.dll "C:\Program Files\Conquer\"
del "C:\Program Files\Conquer\someOtherDll.dll"
....
```
4.You start the batch file. As it has a delay, the JVM has to end in this delay to avoid having open files.
5.The launcher starts (automatically). (And can maybe show an dialog like: Updates are installed!)